### PR TITLE
Twitter

### PR DESCRIPTION
**Pull request by kohanov**

### CODE REVIEW

**Comment by krasnotsvetov** [comment]
See comment here[https://github.com/MentionBattle/SNAdapter/commit/c03ea9a1f50a98d5abdb7541f92bb84c001f9f18]

**Review by krasnotsvetov — CHANGES_REQUESTED** [review]
Above

File: src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/initalizers/TwitterTokens.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package org.mentionbattle.snadapter.impl.socialnetworks.initalizers
+
+import org.mentionbattle.snadapter.api.core.Component
+import org.mentionbattle.snadapter.api.core.SocialNetworkInitializer
+
+@Component
+@SocialNetworkInitializer("TwitterTokens")
+internal class TwitterTokens(val map: Map<String, String>) {
+    val consumerKey= map.get("consumerKey")
```
Comment (krasnotsvetov) [inline src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/initalizers/TwitterTokens.kt]: Use "by map" delegate.
[Example](https://github.com/MentionBattle/SNAdapter/blob/master/src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/initalizers/VkServiceToken.kt)
[Docs](http://kotlinlang.org/docs/reference/delegated-properties.html)

File: src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/handlers/TwitterHandler.kt [hunk 2]
```diff
@@ -1,36 +1,100 @@
...
+import twitter4j.StatusListener
+import twitter4j.FilterQuery
+import twitter4j.TwitterStreamFactory
+import twitter4j.TwitterStream
 
 @SocialNetwork("Twitter")
-internal class TwitterHandler(eventQueue : PrimitiveEventQueue) : SocialNetworkHandler, EventHandler {
+internal class TwitterHandler(token: TwitterTokens, tags: Tags, eventQueue: PrimitiveEventQueue) : SocialNetworkHandler, EventHandler {
 
     val eventQueue = eventQueue
```
Comment (krasnotsvetov) [inline src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/handlers/TwitterHandler.kt]: Add private modifiers to all inner data.
Remove unnessary fields.

File: src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/handlers/TwitterHandler.kt [hunk 3]
```diff
@@ -1,36 +1,100 @@
...
     var isWorking = true
+    val tokens = token
+    val tags = tags
+    lateinit var twitterStream: TwitterStream
+
+    private val listener = object : StatusListener {
+        override fun onStatus(status: Status) {
+            val url = "https://twitter.com/" + status.user.screenName + "/status/" + status.id.toString()
```
Comment (krasnotsvetov) [inline src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/handlers/TwitterHandler.kt]: Use "val" where  it is possible.

File: src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/initalizers/TwitterTokens.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package org.mentionbattle.snadapter.impl.socialnetworks.initalizers
+
+import org.mentionbattle.snadapter.api.core.Component
+import org.mentionbattle.snadapter.api.core.SocialNetworkInitializer
+
+@Component
```
Comment (krasnotsvetov) [inline src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/initalizers/TwitterTokens.kt]: The Component annotation should be removed

File: src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/handlers/TwitterHandler.kt [hunk 5]
```diff
@@ -1,36 +1,100 @@
...
+            ex.printStackTrace()
+        }
+    }
+
+
     override fun handleEvent(event: Event) {
         when (event) {
             is ExitEvent -> {
-                isWorking = false
+                twitterStream.cleanUp(); // shutdown internal stream consuming thread
```
Comment (krasnotsvetov) [inline src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/handlers/TwitterHandler.kt:68]: The resourced will not free, if we crash in processData or we will not shutdown twitterStream if  cleanUp crashes

Comment (novikd) [reply src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/handlers/TwitterHandler.kt:68]: If application crashes, then all the resources we've got will be freed. Am I right? Or I missed something?

**Review by novikd — APPROVED** [review]
Quite good. Needs some small fixes and can be merged into master.

File: src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/handlers/TwitterHandler.kt (hunk 5 above)
Comment (krasnotsvetov) [reply src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/handlers/TwitterHandler.kt:68]: @novikd 
We are launching each handler in seperated coroutines, so if coroutine crashed, the application will continue work and we will lost resources.
More than, I want to a possibility to restart a handler if it crashed, so it significant to free all resources after crash.

**Review by krasnotsvetov — APPROVED** [review]

**Pull request merged**